### PR TITLE
features: append cors plugin

### PR DESCRIPTION
**Pull request by ShiningRush**
### Summary

cors plugin

### Full changelog

- support base cors header
- support multiple origin
- support forcefully match origin and methods with wildcard

### CODE REVIEW

**Comment by moonming** [comment]
@Miss-you please take a look, thx

File: lua/apisix/plugins/cors.lua [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+local plugin_name = "cors"
+local str_find    = string.find
+local str_gmatch  = string.gmatch
+
+local schema = {
+    type = "object",
+    properties = {
+        allow_origins = {
+            description =
+                "you can use '*' to allow all origins when no credentials,"..
```
Comment (membphis) [inline lua/apisix/plugins/cors.lua]: code style: please add a space `credentials," ..`

File: lua/apisix/plugins/cors.lua [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    if not conf.allow_origins then
+        conf.allow_origins = "*"
```
Comment (membphis) [inline lua/apisix/plugins/cors.lua]: we can set default value by `jsonschema`.

File: lua/apisix/plugins/cors.lua [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    end
+
+    if not conf.allow_methods then
+        conf.allow_methods = "*"
```
Comment (membphis) [inline lua/apisix/plugins/cors.lua]: ditto, please fix other similar issues

File: lua/apisix/plugins/cors.lua [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    end
+
+    return true
+end
+
+function _M.header_filter(conf, ctx)
+    if conf.allow_origins == "**" then
+        conf.allow_origins = ngx.var.http_origin or '*'
+    end
+    if str_find(conf.allow_origins, ",") then
```
Comment (membphis) [inline lua/apisix/plugins/cors.lua]: optimize: plain model

https://www.lua.org/manual/5.1/manual.html#pdf-string.find

File: lua/apisix/plugins/cors.lua [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        local finded = false
+        for origin in str_gmatch(conf.allow_origins, "([^,]+)") do
```
Comment (membphis) [inline lua/apisix/plugins/cors.lua]: `ngx.re.gmatch` is better performance

File: t/plugin/cors.t [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+Access-Control-Allow-Origin: *
+Access-Control-Allow-Methods: *
+Access-Control-Allow-Headers: *
+Access-Control-Expose-Headers: *
+Access-Control-Max-Age: 5
+Access-Control-Allow-Credentials: 
+--- no_error_log
+[error]
+
+=== TEST 8: set route(spcific)
```
Comment (membphis) [inline t/plugin/cors.t]: bad test case style

https://github.com/apache/incubator-apisix/blob/master/Contributing.md#check-code-style-and-test-case-style

File: conf/config.yaml [hunk 7]
```diff
@@ -143,7 +143,7 @@ plugins:                          # plugin list
   - proxy-cache
   - tcp-logger
   - proxy-mirror
+  - kafka-logger
```
Comment (membphis) [inline conf/config.yaml]: This commit should be created as a separate PR.

A PR does only one thing.

**Comment by membphis** [comment]
@ShiningRush missing doc

File: doc/plugins/cors-cn.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+- [**测试插件**](#测试插件)
+- [**禁用插件**](#禁用插件)
+
+## 简介
+
+`cors` 插件可以让你轻易地为服务端启用 [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) 的返回头。
+
+## 属性
+
+- `allow_origins`: `可选`.允许跨域访问的Origin, 格式如：`scheme`://`host`:`port`, 比如: https://somehost.com:8081. 多个值使用 `,` 分割, `allow_credential` 为 `false` 时可以使用 `*` 来表示所有 Origin均允许通过。你也可以在启用了 `allow_credential` 后使用 `**` 强制允许所有Origin都通过，但请注意这样存在安全隐患。默认值为 `*`。
```
Comment (membphis) [inline doc/plugins/cors-cn.md]: "`可选`. " -> "`可选`，"

Comment (membphis) [reply doc/plugins/cors-cn.md]: Use Chinese commas

Comment (membphis) [reply doc/plugins/cors-cn.md]: add space between English and Chinese

File: doc/plugins/cors-cn.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+- `allow_methods`: `可选`.允许跨域访问的Method, 比如: `GET`, `POST`等。多个值使用 `,` 分割, `allow_credential` 为 `false` 时可以使用 `*` 来表示所有 Origin均允许通过。你也可以在启用了 `allow_credential` 后使用 `**` 强制允许所有 Method 都通过，但请注意这样存在安全隐患。默认值为 `*`。
+- `allow_headers`: `可选`.允许跨域访问时请求方携带哪些非 CORS规范 以外的 Header,  多个值使用 `,` 分割。默认值为 `*`。
+- `expose_headers`: `可选`.允许跨域访问时响应方携带哪些非 CORS规范 以外的 Header,  多个值使用 `,` 分割。默认值为 `*`。
+- `max_age`: `可选`.浏览器缓存CORS结果的最大时间, 单位为秒, 在这个时间范围内浏览器会复用上一次的检查结果, `-1` 表示不缓存。请注意各个浏览器允许的的最大时间不用, 详情请参考 [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age#Directives). 默认值为 `5`。
+- `allow_credential`: 是否允许跨域访问的请求方携带凭据(如 Cookie 等), 默认值为: `false`。
+
+## 如何启用
+创建 `Route` 或 `Service` 对象，并配置 `cors` 插件。
```
Comment (membphis) [inline doc/plugins/cors-cn.md]: add a blank line

File: doc/plugins/cors-cn.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+```shell
+curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+    "methods": ["GET"],
```
Comment (membphis) [inline doc/plugins/cors-cn.md]: we can remove this line

File: doc/plugins/cors-cn.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+...
+```
+
+## 禁用插件
+
+从配置中移除`cors`插件即可。
+```shell
+$ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+    "methods": ["GET"],
```
Comment (membphis) [inline doc/plugins/cors-cn.md]: we can remove this line

**Comment by membphis** [comment]
add a reference to doc:

https://github.com/apache/incubator-apisix#features

**Comment by ShiningRush** [comment]
The ci show the `./doc/plugins/cors-cn.md` has no license header, but it does have, hmm, what happened to it?

**Comment by membphis** [comment]
you should revert them:

image
